### PR TITLE
Ignore ref-safe-to-escape of unscoped ref to ref struct in MAMM

### LINKED ISSUE #63526: Missing CS8166 when accessing [UnscopedRef] property to assing to an unscoped local
**Version Used**:  4.4.0-2.22419.1 (ae566380)

**Steps to Reproduce**:

1. Paste the following code:
```csharp
ref struct A
{
    [UnscopedRef]
    public Span<byte> Span => default;
}

class Foo
{
    void Bar(ref A a)
    {
        Span<byte> span = a.Span;
    }
}
```

**Expected Behavior**:

The code should produce a CS8166 error (same as #63524, but correctly here), because `span` is not `scoped`.

**Actual Behavior**:

Code compiles fine.

cc. @cston

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to dotnet/csharplang#6450.

This change only relaxes call sites without restricting declaration sites. The other half of the change, which actually introduces the special ReturnOnly escape scope, will come in a subsequent PR.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 1]
```diff
@@ -16814,5 +16778,76 @@ public void M()
...
+
+                ref struct ByteContainer
+                {
+                    public byte B;
+
+                    [UnscopedRef]
+                    public RefByteContainer ByteRef => new RefByteContainer(ref B);
+
+                    [UnscopedRef]
+                    public RefByteContainer GetByteRef() => new RefByteContainer(ref B);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:16806]: I would have tested just a property here but the fix for #63526 has not been merged yet. So only the method has the expected "before" behavior of giving an error at the use site.

**Comment by RikkiGibson** [comment]
@jcouv @cston @dotnet/roslyn-compiler for review

**Comment by cston** [comment]
>         F3(ref r1); // 1

Comment can be removed.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:3147 in 7680d06. [](commit_id = 7680d0632d68f8fb06ab3faad9117a6434de4a02, deletion_comment = False)

**Comment by cston** [comment]
>         y2.F2(ref x2); // 1

Comment can be removed.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:3502 in 7680d06. [](commit_id = 7680d0632d68f8fb06ab3faad9117a6434de4a02, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 2]
```diff
@@ -15058,11 +15025,11 @@ public class A<T>
...
-                // (16,20): error CS8350: This combination of arguments to 'A<int>.F3A(ref R<int>)' is disallowed because it may expose variables referenced by parameter 'r3' outside of their declaration scope
+                // (16,20): error CS8347: Cannot use a result of 'A<int>.F3A(ref R<int>)' in this context because it may expose variables referenced by parameter 'r3' outside of their declaration scope
                 //         return ref F3A(ref r); // 1
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "F3A(ref r)").WithArguments("A<int>.F3A(ref R<int>)", "r3").WithLocation(16, 20),
+                Diagnostic(ErrorCode.ERR_EscapeCall, "F3A(ref r)").WithArguments("A<int>.F3A(ref R<int>)", "r3").WithLocation(16, 20),
                 // (16,28): error CS8168: Cannot return local 'r' by reference because it is not a ref local
-                //         return ref F3A(ref r); // 2
+                //         return ref F3A(ref r); // 1
                 Diagnostic(ErrorCode.ERR_RefReturnLocal, "r").WithArguments("r").WithLocation(16, 28));
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:15034]: >} [](http://example.com/codeflow?start=8&length=1)

I didn't see any tests that verify the inferred scope of a `ref` local. Perhaps add:
```
ref int F3C(ref int x)
{
    var r = new R<int>(ref x);
    ref int y = ref F3A(ref r);
    return ref y; // 1
}
```

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 3]
```diff
@@ -16814,5 +16800,76 @@ public void M()
...
+                        rbc = bc.GetByteRef();
+                    }
+
+                    static RefByteContainer M31([UnscopedRef] ref ByteContainer bc)
+                        // ok. ref-safe-to-escape of 'bc' is 'ReturnOnly'.
+                        => bc.ByteRef;
+
+                    static RefByteContainer M32([UnscopedRef] ref ByteContainer bc)
+                        // ok. ref-safe-to-escape of 'bc' is 'ReturnOnly'.
+                        => bc.GetByteRef();
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:16867]: Can you replicate these tests with a variant where the parameter is plain old `ref` and verify that there is an error?

**Review by jaredpar — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:16867]: Done

**Pull request merged**